# [LAB 05] - PBT 데이터 품질 점검 - 1. Boston Husing Dataset

## #01. 준비 작업 

### 1. 필요한 라이브러리 참조 

In [30]:
import numpy as np
from jussam import load_data
from pandas import DataFrame, read_excel

### 2. 데이터 불러오기 

In [31]:
origin = load_data("boston")
origin.head()

📚 보스턴 인근 지역의 인구·환경·교육·범죄·세금 정보 등을 기반으로 주택의 중위가격(MEDV)을 예측하기 위해 수집된 대표적인 회귀 분석용 주택시장 데이터 (출처: https://www.kaggle.com/datasets/altavish/boston-housing-dataset)

    field    description
--  -------  ---------------------------------------------------------------------------------------------
 0  CRIM     타운별 1인당 범죄율 (Per capita crime rate by town)
 1  ZN       25,000 sqft 이상 주거용 대지 비율 (Proportion of residential land zoned for large lots)
 2  INDUS    비소매 상업구역 비율 (Proportion of non-retail business acres per town)
 3  CHAS     명목형, 찰스강 인접 여부 (Charles River dummy variable: 1 if tract bounds river; 0 otherwise)
 4  NOX      대기 오염 정도 (Nitric oxides concentration, parts per 10 million)
 5  RM       주택당 평균 방(Room) 수 (Average number of rooms per dwelling)
 6  AGE      1940년 이전 건축 비율 (Proportion of owner-occupied units built prior to 1940)
 7  DIS      5개 주요 고용센터까지의 거리 (Weighted distances to five Boston employment centers)
 8  RAD      고속도로 접근성 지수 (Index of accessibility to radial highways)
 9  T

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.006,18.000,2.310,0,0.538,6.575,65.200,4.090,1,296,15.300,396.900,4.980,24.000
1,0.027,0.000,7.070,0,0.469,6.421,78.900,4.967,2,242,17.800,396.900,9.140,21.600
2,0.027,0.000,7.070,0,0.469,7.185,61.100,4.967,2,242,17.800,392.830,4.030,34.700
3,0.032,0.000,2.180,0,0.458,6.998,45.800,6.062,3,222,18.700,394.630,2.940,33.400
4,0.069,0.000,2.180,0,0.458,7.147,54.200,6.062,3,222,18.700,396.900,5.330,36.200


## #02. 자료형 점검 

### 1. 자료형 확인 

In [32]:
origin.info()

<class 'pandas.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


### 2. 자료형 변환 
- CHAS는 명목형이므로 category타입으로 변환 

In [33]:
# 원본 데이터 프레임 복사 
df1 = origin.copy()
df1["CHAS"] = df1["CHAS"].astype("category")
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   CRIM     506 non-null    float64 
 1   ZN       506 non-null    float64 
 2   INDUS    506 non-null    float64 
 3   CHAS     506 non-null    category
 4   NOX      506 non-null    float64 
 5   RM       506 non-null    float64 
 6   AGE      506 non-null    float64 
 7   DIS      506 non-null    float64 
 8   RAD      506 non-null    int64   
 9   TAX      506 non-null    int64   
 10  PTRATIO  506 non-null    float64 
 11  B        506 non-null    float64 
 12  LSTAT    506 non-null    float64 
 13  MEDV     506 non-null    float64 
dtypes: category(1), float64(11), int64(2)
memory usage: 52.0 KB


## #03. 데이터 중복 점검 

### 1. 행 단위 완전 중복 확인 
- 중복된 행에 대하여 True로 표시하고, 중복되지 않은 행은 False로 표시한다. 

In [34]:
dup = df1.duplicated()
dup

0      False
1      False
2      False
3      False
4      False
       ...  
501    False
502    False
503    False
504    False
505    False
Length: 506, dtype: bool

In [35]:
# 중복 데이터 개수
dup.sum()

np.int64(0)

### 2. 중복 데이터 제거 
- 여기서는 필요없는 처리이지만 코드 학습을 위해 수행 

In [36]:
df2 = df1.drop_duplicates()
df2.duplicated().sum()

np.int64(0)

## #03. 형식, 표기, 단위 점검 

### 1. 명목형 변수의 단위 확인 
- 명목형 변수는 값의 종류별 분포를 확인하는 것으로 단위를 확인할 수 있다. 

In [37]:
df2['CHAS'].value_counts()

CHAS
0    471
1     35
Name: count, dtype: int64

### 2. 연속형 변수의 단위 확인을 위한 변수명 추출 

In [38]:
# 값을 검사 수치형 변수의 이름 추출 
fields = df2.select_dtypes(include="number").columns.to_list()
print(fields)

['CRIM', 'ZN', 'INDUS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']


### 3. 연속형 변수의 단위 확인 
- 값의 구간(최소, 최대값)을 확인하여 데이터 명세와 비교해 본다. 

In [39]:
minmax = [] # 최소값과 최대값을 저장할 리스트 

for field in fields:
    min_value = df2[field].min()
    max_value = df2[field].max()
    minmax.append({"min":min_value, "max":max_value})


# 최소값과 최대값을 데이터 프레임으로 변환하여 출력 
minmax_df = DataFrame(minmax, index=fields)
minmax_df

,min,max
CRIM,0.006,88.976
ZN,0.000,100.000
INDUS,0.460,27.740
NOX,0.385,0.871
RM,3.561,8.780
AGE,2.900,100.000
DIS,1.130,12.127
RAD,1.000,24.000
TAX,187.000,711.000
PTRATIO,12.600,22.000


## #05. 결측치 점검 

### 1. 결측치 확인 

In [40]:
na_count = df2.isna().sum()
na_count

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

### 2. 결측치 비율 계산을 위한 데이터 크기 확인 

In [41]:
# 데이터의 행과 열의 수 확인 
rows, cols = df2.shape
print(f"rows: {rows}, cols: {cols}")

rows: 506, cols: 14


### 3. 결측치 비율 확인
- 각 변수의 결측치 수를 전체 데이터 행 수로 나눈다. 
- 결측치 비율이 5%(0.05) 미만이므로 전체 결과에 미치는 영향이 미미 하므로 결측치 삭제 가능 

In [42]:
na_ratio = na_count / rows 
na_ratio

CRIM      0.000
ZN        0.000
INDUS     0.000
CHAS      0.000
NOX       0.000
RM        0.000
AGE       0.000
DIS       0.000
RAD       0.000
TAX       0.000
PTRATIO   0.000
B         0.000
LSTAT     0.000
MEDV      0.000
dtype: float64

## #06. 품질 점검 결과 저장 

### 1. 엑셀 파일로 저장하기 
- 다음 분석 과정에서 이어서 진행할 수 있도록 품질 점검이 완료된 데이터 셋을 저장한다. 

In [43]:
df2.to_excel("boston_qtcheck.xlsx", index=False)

## #07. 기술 통계량 확인 준비 작업 

### 1. 데이터 불러오기 
- 실제 분석 과정에서는 새로운 jupyter파일을 생성하고 \
앞 단계에서 저장한 데이터를 불러와서 진행하는 것을 권장 

In [44]:
# origin_qt = read_excel("boston_qtcheck.xlsx")

# 수업용 자료이므로 라이브러리를 통해서도 로드 가능함 
origin_qt = load_data("boston_qtcheck")

📚 보스턴 주택 가격 데이터셋의 품질 검사 완료 버전 (출처: 자체 정제)

    field    description
--  -------  ---------------------------------------------------------------------------------------------
 0  CRIM     타운별 1인당 범죄율 (Per capita crime rate by town)
 1  ZN       25,000 sqft 이상 주거용 대지 비율 (Proportion of residential land zoned for large lots)
 2  INDUS    비소매 상업구역 비율 (Proportion of non-retail business acres per town)
 3  CHAS     명목형, 찰스강 인접 여부 (Charles River dummy variable: 1 if tract bounds river; 0 otherwise)
 4  NOX      대기 오염 정도 (Nitric oxides concentration, parts per 10 million)
 5  RM       주택당 평균 방(Room) 수 (Average number of rooms per dwelling)
 6  AGE      1940년 이전 건축 비율 (Proportion of owner-occupied units built prior to 1940)
 7  DIS      5개 주요 고용센터까지의 거리 (Weighted distances to five Boston employment centers)
 8  RAD      고속도로 접근성 지수 (Index of accessibility to radial highways)
 9  TAX      재산세율 (Full-value property-tax rate per $10,000)
10  PTRATIO  교육구 학생-교사 비율 (Pupil–teacher ratio by town)
11  

## #01. 준비작업 

### 2. 타입 변환 
- category 타입으로의 변환은 데이터를 불러올 때 마다 수행해야 한다. 

In [45]:
df3 = origin_qt.copy()
df3['CHAS'] = df3['CHAS'].astype('category')
df3.info()

<class 'pandas.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   CRIM     506 non-null    float64 
 1   ZN       506 non-null    float64 
 2   INDUS    506 non-null    float64 
 3   CHAS     506 non-null    category
 4   NOX      506 non-null    float64 
 5   RM       506 non-null    float64 
 6   AGE      506 non-null    float64 
 7   DIS      506 non-null    float64 
 8   RAD      506 non-null    int64   
 9   TAX      506 non-null    int64   
 10  PTRATIO  506 non-null    float64 
 11  B        506 non-null    float64 
 12  LSTAT    506 non-null    float64 
 13  MEDV     506 non-null    float64 
dtypes: category(1), float64(11), int64(2)
memory usage: 52.0 KB


## #02. 기술 통계량 표 생성 

### 1. 연속형 변수의 기술 통계량 표 
- 위도와 경도는 연속형 변수가 아니므로 제외하고 기술 통계량을 확인해야 한다. 

In [46]:
# T는 데이터 프레임의 행과 열을 바꿔주는 역할 
desc_df = df3.describe().T
desc_df

,count,mean,std,min,25%,50%,75%,max
CRIM,506.000,3.614,8.602,0.006,0.082,0.257,3.677,88.976
ZN,506.000,11.364,23.322,0.000,0.000,0.000,12.500,100.000
INDUS,506.000,11.137,6.860,0.460,5.190,9.690,18.100,27.740
NOX,506.000,0.555,0.116,0.385,0.449,0.538,0.624,0.871
RM,506.000,6.285,0.703,3.561,5.885,6.208,6.623,8.780
AGE,506.000,68.575,28.149,2.900,45.025,77.500,94.075,100.000
DIS,506.000,3.795,2.106,1.130,2.100,3.207,5.188,12.127
RAD,506.000,9.549,8.707,1.000,4.000,5.000,24.000,24.000
TAX,506.000,408.237,168.537,187.000,279.000,330.000,666.000,711.000
PTRATIO,506.000,18.456,2.165,12.600,17.400,19.050,20.200,22.000


### 2. 명목형 변수의 기술 통계량 표 
- 데이터의 수(count), 값의 종류 수(unique), 최빈값(top), 최빈값의 수(freq)를 알 수 있다. 


In [47]:
cate_desc_df = df3.describe(include='category').T
cate_desc_df

,count,unique,top,freq
CHAS,506,2,0,471


### 3. 명목형 변수의 기술 통계량 
- 명목형 변수는 값의 종류별로 데이터 수와 비율을 확인할 수 있다. 

In [48]:
# 명목형 변수의 필드명 추출 
cate_fields = df3.select_dtypes(include='category').columns

# 명목형 변수의 고유값과 빈도수 계산 
for field in cate_fields: 
    # 값의 종류별 개수 
    vcount = df3[field].value_counts()
    # 전체 데이터 길이 대비 비율
    percent = vcount / df3.shape[0] 

    # 결과 저장 
    df = DataFrame({'count': vcount, 'percent': percent})
    # 결과 출력 
    display(df)

,count,percent
CHAS,,
0,471,0.931
1,35,0.069


## #03. 중심 수준 파악 

### 1. 평균-중앙값의 상대 차이율 계산 

In [49]:
# "평균-중앙값 상대 차이율 = |평균 -중앙값| / 중앙값" 컬럼 추가 
desc_df['rel_diff'] = abs(desc_df['mean'] - desc_df['50%'] ) / desc_df['50%']

# 상대 차이율 의미 컬럼 추가 
conditions = [desc_df['rel_diff'] < 0.1, desc_df['rel_diff'] < 0.5]
choices = ['similar', 'diff']
desc_df['rdiff_flag'] = np.select(conditions, choices, default='large_diff')

# 필요한 컬럼만 선택하여 출력 


## #04. 이상치 신호 감지 

### 1. IQR 이상치 경계값 계산 

In [50]:
# iqr 
desc_df['iqr'] = desc_df['75%'] - desc_df['25%']    

# 상한 이상치 경계 
desc_df['upper_bound'] = desc_df['75%'] + 1.5 * desc_df['iqr']

# 하한 이상치 경계 
desc_df['lower_bound'] = desc_df['25%'] - 1.5 * desc_df['iqr']

# 필요한 컬럼만 선택하여 출력 
desc_df[['iqr', 'upper_bound', 'lower_bound']]

,iqr,upper_bound,lower_bound
CRIM,3.595,9.070,-5.311
ZN,12.500,31.250,-18.750
INDUS,12.910,37.465,-14.175
NOX,0.175,0.886,0.187
RM,0.738,7.731,4.778
AGE,49.050,167.650,-28.550
DIS,3.088,9.821,-2.532
RAD,20.000,54.000,-26.000
TAX,387.000,1246.500,-301.500
PTRATIO,2.800,24.400,13.200


### 2. 명목형 변수를 제외한 데이터 프레임 
- 이상치는 연속형 변수에 대한 개념이므로 명목형 변수를 제외한 데이터 프레임 생성 

In [51]:
# 명목형 변수의 필드명 추출 
cate_fields = df3.select_dtypes(include='category').columns

df4 = df3.drop(columns=cate_fields)
df4.head()

,CRIM,ZN,INDUS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.006,18.000,2.310,0.538,6.575,65.200,4.090,1,296,15.300,396.900,4.980,24.000
1,0.027,0.000,7.070,0.469,6.421,78.900,4.967,2,242,17.800,396.900,9.140,21.600
2,0.027,0.000,7.070,0.469,7.185,61.100,4.967,2,242,17.800,392.830,4.030,34.700
3,0.032,0.000,2.180,0.458,6.998,45.800,6.062,3,222,18.700,394.630,2.940,33.400
4,0.069,0.000,2.180,0.458,7.147,54.200,6.062,3,222,18.700,396.900,5.330,36.200


### 3. 상한 이상치 탐지 

In [52]:
# 상한 이상치 수 
desc_df['upper_outliers'] = ((df4 > desc_df['upper_bound'])).sum()

# 상한 이상치 수 비율 
desc_df['upper_outliers_ratio'] = desc_df['upper_outliers'] / df4.shape[0]

# 필요한 컬럼만 선택하여 출력 
desc_df[['upper_outliers', 'upper_outliers_ratio']]

,upper_outliers,upper_outliers_ratio
CRIM,66,0.130
ZN,68,0.134
INDUS,0,0.000
NOX,0,0.000
RM,22,0.043
AGE,0,0.000
DIS,5,0.010
RAD,0,0.000
TAX,0,0.000
PTRATIO,0,0.000


### 4. 하한 이상치 탐지 

In [53]:
# 하한 이상치 수 
desc_df['lower_outliers'] = (df4 < desc_df['lower_bound']).sum()

# 하한 이상치 수 비율
desc_df['lower_outliers_ratio'] = desc_df['lower_outliers'] / df4.shape[0]  

# 필요한 컬럼만 선택하여 출력 
desc_df[['lower_outliers', 'lower_outliers_ratio']]

,lower_outliers,lower_outliers_ratio
CRIM,0,0.000
ZN,0,0.000
INDUS,0,0.000
NOX,0,0.000
RM,8,0.016
AGE,0,0.000
DIS,0,0.000
RAD,0,0.000
TAX,0,0.000
PTRATIO,15,0.030


### 5. 전체 이상치 집계 

In [54]:
# 통합 이상치 수 
desc_df['outliers'] = desc_df['upper_outliers'] + desc_df['lower_outliers']

# 통합 이상치 수 비율 
desc_df['outliers_ratio'] = desc_df['outliers']/ df3.shape[0] 

# 이상치 탐지 결과 확인 
desc_df[['upper_outliers', 'upper_outliers_ratio', 'lower_outliers', 'lower_outliers_ratio', 'outliers', 'outliers_ratio']]

,upper_outliers,upper_outliers_ratio,lower_outliers,lower_outliers_ratio,outliers,outliers_ratio
CRIM,66,0.130,0,0.000,66,0.130
ZN,68,0.134,0,0.000,68,0.134
INDUS,0,0.000,0,0.000,0,0.000
NOX,0,0.000,0,0.000,0,0.000
RM,22,0.043,8,0.016,30,0.059
AGE,0,0.000,0,0.000,0,0.000
DIS,5,0.010,0,0.000,5,0.010
RAD,0,0.000,0,0.000,0,0.000
TAX,0,0.000,0,0.000,0,0.000
PTRATIO,0,0.000,15,0.030,15,0.030


## #05. 비대칭 신호 확인 

### 1. 왜도 점검 

In [55]:
# 왜도 계산 
desc_df['skew'] = df4.skew()

# 왜도를 통한 분포 형태 해석 
conditions_skew = [(desc_df['skew'] < -0.5), (desc_df['skew'] > 0.5)]
choices_skew = ['lesf tail', 'right tail']
desc_df['skew_interpret'] = np.select(conditions_skew, choices_skew, default='symmetric')

# 필요한 컬럼만 선택하여 출력 
desc_df[['skew', 'skew_interpret']]

,skew,skew_interpret
CRIM,5.223,right tail
ZN,2.226,right tail
INDUS,0.295,symmetric
NOX,0.729,right tail
RM,0.404,symmetric
AGE,-0.599,lesf tail
DIS,1.012,right tail
RAD,1.005,right tail
TAX,0.670,right tail
PTRATIO,-0.802,lesf tail


### 2. 첨도 점검 


In [59]:
desc_df['kurt'] = df4.kurt()

# 첨도를 통한 분포 형태 해석 
conditions_kurt = [(desc_df['kurt'] < 0), (desc_df['kurt'] > 0)]
choices_kurt = ['platykurtic', 'leptokurtic']
desc_df['kurt_interpret'] = np.select(conditions_kurt, choices_kurt, default='mesokutic')

# 필요한 컬럼만 선택하여 출력 
desc_df[['kurt', 'kurt_interpret']]

,kurt,kurt_interpret
CRIM,37.131,leptokurtic
ZN,4.032,leptokurtic
INDUS,-1.234,platykurtic
NOX,-0.065,platykurtic
RM,1.892,leptokurtic
AGE,-0.968,platykurtic
DIS,0.488,leptokurtic
RAD,-0.867,platykurtic
TAX,-1.142,platykurtic
PTRATIO,-0.285,platykurtic


### 3. 로그 변환 필요성 판단 함수 정의 

In [ ]:
def judge_log_transfrom(skew, kurt): 
    if skew >= 1:                 # 강한 우측 꼬리 분포 
        return "log1p"
    elif skew > 0.5 and kurt > 0: # 우측 꼬리 분포이면서 첨도가 높은 경우 
        return "log1p"
    elif skew <= -1:              # 강한 좌측 꼬리 분포 
        return "reverse_log1p"
    elif skew < -0.5 and kurt > 0: # 좌측 꼬리 분포이면서 첨도가 높은 경우 
        return "reverse_log1p"
    else: 
        return "none"

### 4. 로그 변환 필요성 판정 

In [63]:
desc_df['log_need'] = desc_df.apply(lambda row: judge_log_transfrom(
    row['skew'], row['kurt']), axis=1)

# 로그 변환 필요성 결과 확인 
desc_df[['skew', 'kurt', 'log_need']]

,skew,kurt,log_need
CRIM,5.223,37.131,log1p
ZN,2.226,4.032,log1p
INDUS,0.295,-1.234,none
NOX,0.729,-0.065,none
RM,0.404,1.892,none
AGE,-0.599,-0.968,none
DIS,1.012,0.488,log1p
RAD,1.005,-0.867,log1p
TAX,0.670,-1.142,none
PTRATIO,-0.802,-0.285,none


## #06. 최종 기술 통계량 확인 

### 2. 기술 통계량 표 저장 

In [64]:
desc_df.to_excel("boston_qtcheck_desc.xlsx")

In [65]:
# 연습문제: 보험료 가격 결정 구조 분석을 위한 데이터셋의 품질 검사
# 배경
# 왜 어떤 사람의 의료보험 청구 비용은 높고, 어떤 사람은 낮을까? (의료비를 좌우하는 진짜 요인은 무엇일까?)
# 의료 비용은 개인 건강 상태뿐 아니라 나이, 성별, 거주지역, 생활습관, 흡연 여부에 따라 체계적으로 달라집니다.
# KEY QUESTION: 흡연이 의료비를 얼마나 바꿀까?

# 흡연 여부는 의료비를 얼마나 크게 변화시킬까?
# 젊은 사람도 흡연자라면 높은 비용을 지불할까?


# insurance 데이터셋으로 다음 3가지를 수행하세요
# 01 데이터 개요 확인

# 다음 페이지의 Insurance 데이터셋 구조를 먼저 살펴봅니다. (1,338 × 7 / 관측치 × 변수)
# 02 데이터 품질 검사

# 결측치·중복·이상치·자료형을 점검해 분석 가능 상태로 정리합니다. (Quality Check)
# 03 기술 통계량 확인

# 평균·분포·범주 빈도를 통해 데이터의 형태를 파악합니다. (Describe)

# 데이터셋 정보
# 개인 의료보험료 데이터 / 관측치 수: 1,338개 / 변수 수: 7개

# 분석 대상·단위: 개인 의료보험 가입자 (1,338명), 1행 = 1인의 의료보험 기록
# 출처: Kaggle – Medical Cost Personal Datasets (mirichoi0218)
# 공간: 미국 4개 지역 (northeast, northwest, southeast, southwest)

# 유형분류변수설명값 범위/구분종속변수의료비charges개인에게 청구된 연간 의료보험료 (USD)약 1,122 ~ 63,770독립변수인구 통계age피보험자의 나이 (세)18 ~ 64독립변수인구 통계sex성별 (범주형)female, male독립변수건강 지표bmi체질량지수 kg/m² (이상 18.5~24.9)약 15.96 ~ 53.13독립변수건강 지표smoker흡연 여부 (범주형)yes, no독립변수가족 구성children보험에 등재된 부양 자녀 수0 ~ 5독립변수지역region미국 거주 지역 (범주형)NE, NW, SE, SW

# [LAB-05] 1. 연습문제 - 보험료 가격 결정 구조 데이터셋 품질 검사

## #01. 준비작업

### 1. 필요한 라이브러리 참조

In [ ]:
import numpy as np
from jussam import load_data
from pandas import DataFrame, read_excel

📦 연세대학교 주영아 교수가 제작한 라이브러리를 사용중입니다.
📧 Email(1): j.purplerose@yonsei.ac.kr
📧 Email(2): j.purplerose@gmail.com
📝 Website: https://juyounga.kr/


### 2. 데이터 불러오기

In [ ]:
origin = load_data("insurance")
origin.head()

📚 개인의 나이·성별·BMI·흡연 여부·거주 지역 등 기본 건강·인구학   적 정보를 바탕으로 의료보험 청구 비용(charges)을 예측하도록 구성된, 선형회귀와 머신러닝 실습에 널리 사용되는 대표적인 보험 비용 데이터셋 (출처: https://www.kaggle.com/datasets/mirichoi0218/insurance)

    field     description
--  --------  --------------------------------------------------------------
 0  age       가입자의 나이(세)
 1  sex       성별 (male, female)
 2  bmi       체질량 지수(Body Mass Index)
 3  children  부양 자녀 수(보험 내 자녀 수)
 4  smoker    흡연 여부 (yes / no)
 5  region    미국 내 거주 지역 (northeast, northwest, southeast, southwest)
 6  charges   의료보험 청구 비용(달러). 예측해야 하는 타깃 변수.



,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.924
1,18,male,33.770,1,no,southeast,1725.552
2,28,male,33.000,3,no,southeast,4449.462
3,33,male,22.705,0,no,northwest,21984.471
4,32,male,28.880,0,no,northwest,3866.855


## #02. 자료형 점검

### 1. 자료형 확인

In [ ]:
origin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


### 2. 자료형 변환

In [ ]:
# 원본 데이터 프레임 복사
df1 = origin.copy()
df1["sex"] = df1["sex"].astype("category")
df1["smoker"] = df1["smoker"].astype("category")
df1["region"] = df1["region"].astype("category")
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   age       1338 non-null   int64   
 1   sex       1338 non-null   category
 2   bmi       1338 non-null   float64 
 3   children  1338 non-null   int64   
 4   smoker    1338 non-null   category
 5   region    1338 non-null   category
 6   charges   1338 non-null   float64 
dtypes: category(3), float64(2), int64(2)
memory usage: 46.3 KB


## #03. 데이터 중복 점검

### 1. 행단위 완전 중복 확인

In [ ]:
dup = df1.duplicated()
dup

0       False
1       False
2       False
3       False
4       False
        ...  
1333    False
1334    False
1335    False
1336    False
1337    False
Length: 1338, dtype: bool

In [ ]:
# 중복 데이터 개수
dup.sum()

np.int64(1)

> 앗! 중복 데이터가 발견되었다.

### 2. 중복 데이터 제거

In [ ]:
df2 = df1.drop_duplicates()

# 중복 제거 후 중복 개수 다시 확인
df2.duplicated().sum()

np.int64(0)

## #04. 형식·표기·단위 점검

### 1. 명목형 변수의 단위 확인

In [ ]:
fields = df2.select_dtypes(include=["category"]).columns

for field in fields:
    display(df2[field].value_counts())

sex
male      675
female    662
Name: count, dtype: int64

smoker
no     1063
yes     274
Name: count, dtype: int64

region
southeast    364
southwest    325
northeast    324
northwest    324
Name: count, dtype: int64

> 데이터의 종류에는 문제가 없으나, 라벨링 처리는 필요하다.

### 2. 연속형 변수의 단위 확인을 위한 변수명 추출

In [ ]:
# 값을 검사 수치형 변수의 이름 추출
fields = df2.select_dtypes(include="number").columns.to_list()
print(fields)

['age', 'bmi', 'children', 'charges']


### 3. 연속형 변수의 단위 확인

In [ ]:
minmax = [] # 최소값과 최대값을 저장할 리스트

for field in fields:
    min_value = df2[field].min()
    max_value = df2[field].max()
    minmax.append({"min": min_value, "max": max_value})

# 최소값과 최대값을 데이터 프레임으로 변환하여 출력
minmax_df = DataFrame(minmax, index=fields)
minmax_df

,min,max
age,18.000,64.000
bmi,15.960,53.130
children,0.000,5.000
charges,1121.874,63770.428


> 데이터 명세의 내용에서 크게 벗어나는 범위는 없는것으로 판단됨

## #05. 결측치 점검

### 1. 결측치 확인

In [ ]:
na_count = df2.isna().sum()
na_count

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

### 2. 결측치 비율 계산을 위한 데이터 크기 확인

In [ ]:
# 데이터의 행과 열의 수 확인
rows, cols = df2.shape
print(f"rows: {rows}, cols: {cols}")

rows: 1337, cols: 7


### 3. 결측치 비율 확인

In [ ]:
na_ratio = na_count / rows
na_ratio

age        0.000
sex        0.000
bmi        0.000
children   0.000
smoker     0.000
region     0.000
charges    0.000
dtype: float64

## #06. 품질 점검 결과 데이터셋 저장

In [ ]:
df2.to_excel("insurance_qtcheck.xlsx", index=False)

## #07. 기술 통계량 확인 준비작업

### 1. 데이터 불러오기

In [ ]:
#origin_qt = read_excel("insurance_qtcheck.xlsx")

# 수업용 자료이므로 라이브러리를 통해서도 로드 가능함
origin_qt = load_data("insurance_qtcheck")

### 2. 타입변환

- category 타입으로의 변환은 데이터를 불러올 때 마다 진행해야 한다.

In [ ]:
df3 = origin_qt.copy()
df3["sex"] = df3["sex"].astype("category")
df3["smoker"] = df3["smoker"].astype("category")
df3["region"] = df3["region"].astype("category")
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1337 entries, 0 to 1336
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   age       1337 non-null   int64   
 1   sex       1337 non-null   category
 2   bmi       1337 non-null   float64 
 3   children  1337 non-null   int64   
 4   smoker    1337 non-null   category
 5   region    1337 non-null   category
 6   charges   1337 non-null   float64 
dtypes: category(3), float64(2), int64(2)
memory usage: 46.3 KB


## #08. 기술 통계량 표 생성

### 1. 연속형 변수의 기술 통계량 표

- 명목형 변수는 자동으로 제외된다.

In [ ]:
# T는 데이터프레임의 행과 열을 바꿔주는 역할
desc_df = df3.describe().T
desc_df

,count,mean,std,min,25%,50%,75%,max
age,1337.000,39.222,14.044,18.000,27.000,39.000,51.000,64.000
bmi,1337.000,30.663,6.100,15.960,26.290,30.400,34.700,53.130
children,1337.000,1.096,1.206,0.000,0.000,1.000,2.000,5.000
charges,1337.000,13279.121,12110.360,1121.874,4746.344,9386.161,16657.717,63770.428


### 2. 명목형 변수의 기술 통계량 표

- 데이터의 수(count), 값의 종류 수(unique), 최빈값(top), 최빈값의 수(freq)를 알 수 있다.

In [ ]:
cate_desc_df = df3.describe(include='category').T
cate_desc_df

,count,unique,top,freq
sex,1337,2,male,675
smoker,1337,2,no,1063
region,1337,4,southeast,364


### 3. 명목형 변수 기술 통계량

- 명목형 변수는 값의 종류별로 데이터 수와 비율을 확인할 수 있다.

In [ ]:
# 명목형 변수의 필드명 추출
cate_fields = df3.select_dtypes(include='category').columns

# 명목형 변수의 고유값과 빈도수 계산
for field in cate_fields:
    # 값의 종류별 개수
    vcount = df3[field].value_counts()
    # 전체 데이터 길이 대비 비율
    percent = vcount / df3.shape[0]

    # 결과 저장
    df = DataFrame({'count': vcount, 'percent': percent})
    # 결과 출력
    display(df)

,count,percent
sex,,
male,675,0.505
female,662,0.495


,count,percent
smoker,,
no,1063,0.795
yes,274,0.205


,count,percent
region,,
southeast,364,0.272
southwest,325,0.243
northeast,324,0.242
northwest,324,0.242


## #09. 중심 수준 파악

### 1. **평균-중앙값의 상대 차이율** 계산

- 평균-중앙값 상대 차이율 = |평균 - 중앙값| / 중앙값

| 구간 | 의미 |
|--------|-----------------|
| 0 ~ 0.1   | 평균과 중앙값이 거의 비슷함 |
| 0.1 ~ 0.5  | 약간 차이 있음        |
| 0.5 | 차이 큼, 왜도·극단값 의심 |

In [ ]:
# "평균-중앙값 상대 차이율 = |평균 - 중앙값| / 중앙값" 컬럼 추가
desc_df['rel_diff'] = abs(desc_df['mean'] - desc_df['50%']) / desc_df['50%']

# 상대 차이율 의미 컬럼 추가
conditions = [desc_df['rel_diff'] < 0.1, desc_df['rel_diff'] < 0.5]
choices = ['similar', 'diff']
desc_df['rdiff_flag'] = np.select(conditions, choices, default='large_diff')

# 필요한 컬럼만 선택하여 출력
desc_df[['mean', '50%', 'rel_diff', 'rdiff_flag']]

,mean,50%,rel_diff,rdiff_flag
age,39.222,39.000,0.006,similar
bmi,30.663,30.400,0.009,similar
children,1.096,1.000,0.096,similar
charges,13279.121,9386.161,0.415,diff


## #10. 이상치 신호 탐지

### 1. IQR, 이상치 경계값 계산

In [ ]:
# iqr
desc_df['iqr'] = desc_df['75%'] - desc_df['25%']

# 상한 이상치 경계
desc_df['upper_bound'] = desc_df['75%'] + 1.5 * desc_df['iqr']

# 하한 이상치 경계
desc_df['lower_bound'] = desc_df['25%'] - 1.5 * desc_df['iqr']

# 필요한 컬럼만 선택하여 출력
desc_df[['iqr', 'upper_bound', 'lower_bound']]

,iqr,upper_bound,lower_bound
age,24.000,87.000,-9.000
bmi,8.410,47.315,13.675
children,2.000,5.000,-3.000
charges,11911.373,34524.778,-13120.716


### 2. 명목형 변수를 제외한 데이터 프레임

- 이상치는 연속형 변수에 대한 개념이므로 명목형 변수를 제외한 데이터프레임 생성

In [ ]:
# 명목형 변수의 필드명 추출
cate_fields = df3.select_dtypes(include='category').columns

df4 = df3.drop(columns=cate_fields)
df4.head()

,age,bmi,children,charges
0,19,27.900,0,16884.924
1,18,33.770,1,1725.552
2,28,33.000,3,4449.462
3,33,22.705,0,21984.471
4,32,28.880,0,3866.855


### 3. 상한 이상치 탐지

In [ ]:
# 상한 이상치 수
desc_df['upper_outliers'] = ((df4 > desc_df['upper_bound'])).sum()

# 상한 이상치 수 비율
desc_df['upper_outliers_ratio'] = desc_df['upper_outliers'] / df4.shape[0]

# 필요한 컬럼만 선택하여 출력
desc_df[['upper_outliers', 'upper_outliers_ratio']]

,upper_outliers,upper_outliers_ratio
age,0,0.000
bmi,9,0.007
children,0,0.000
charges,139,0.104


### 4. 하한 이상치 탐지

In [ ]:
# 하한 이상치 수
desc_df['lower_outliers'] = (df4 < desc_df['lower_bound']).sum()

# 하한 이상치 수 비율
desc_df['lower_outliers_ratio'] = desc_df['lower_outliers'] / df4.shape[0]

# 필요한 컬럼만 선택하여 출력
desc_df[['lower_outliers', 'lower_outliers_ratio']]

,lower_outliers,lower_outliers_ratio
age,0,0.000
bmi,0,0.000
children,0,0.000
charges,0,0.000


### 5. 전체 이상치 집계

In [ ]:
# 통합 이상치 수
desc_df['outliers'] = desc_df['upper_outliers'] + desc_df['lower_outliers']

# 통합 이상치 수 비율
desc_df['outliers_ratio'] = desc_df['outliers'] / df4.shape[0]

# 이상치 탐지 결과 확인
desc_df[['upper_outliers', 'upper_outliers_ratio', 
         'lower_outliers', 'lower_outliers_ratio', 
         'outliers', 'outliers_ratio']]

,upper_outliers,upper_outliers_ratio,lower_outliers,lower_outliers_ratio,outliers,outliers_ratio
age,0,0.000,0,0.000,0,0.000
bmi,9,0.007,0,0.000,9,0.007
children,0,0.000,0,0.000,0,0.000
charges,139,0.104,0,0.000,139,0.104


## #11. 비대칭 신호 확인

### 1. 왜도 점검

In [ ]:
# 왜도 계산
desc_df['skew'] = df4.skew()

# 왜도를 통한 분포 형태 해석
conditions_skew = [(desc_df['skew'] < -0.5), (desc_df['skew'] > 0.5)]
choices_skew = ['left tail', 'right tail']
desc_df['skew_interpret'] = np.select(conditions_skew, choices_skew,
                                    default='symmetric')

# 필요한 컬럼만 선택하여 출력
desc_df[['skew', 'skew_interpret']]

,skew,skew_interpret
age,0.055,symmetric
bmi,0.284,symmetric
children,0.937,right tail
charges,1.515,right tail


### 2. 첨도 점검

In [ ]:
# 첨도 계산
desc_df['kurt'] = df4.kurt()

# 첨도를 통한 분포 형태 해석
conditions_kurt = [(desc_df['kurt'] < 0), (desc_df['kurt'] > 0)]
choices_kurt = ['platykurtic', 'leptokurtic']
desc_df['kurt_interpret'] = np.select(conditions_kurt, choices_kurt,
                                    default='mesokurtic')

# 필요한 컬럼만 선택하여 출력
desc_df[['kurt', 'kurt_interpret']]

,kurt,kurt_interpret
age,-1.244,platykurtic
bmi,-0.053,platykurtic
children,0.201,leptokurtic
charges,1.604,leptokurtic


### 3. 로그 변환 필요성 판단 함수 정의

In [ ]:
def judge_log_transform(skew, kurt):
    if skew >= 1:                        # 강한 우측 꼬리 분포
        return "log1p"
    elif skew > 0.5 and kurt > 0:       # 우측 꼬리 분포이면서 첨도가 높은 경우
        return "log1p"
    elif skew <= -1:                    # 강한 좌측 꼬리 분포
        return "reverse_log1p"
    elif skew < -0.5 and kurt > 0:      # 좌측 꼬리 분포이면서 첨도가 높은 경우
        return "reverse_log1p"
    else:                               # 대칭 분포
        return "none"

### 4. 로그 변환 필요성 판정

In [ ]:
desc_df['log_need'] = desc_df.apply(lambda row: judge_log_transform( row['skew'], row['kurt']), axis=1)

# 로그 변환 필요성 결과 확인
desc_df[['skew', 'kurt', 'log_need']]

,skew,kurt,log_need
age,0.055,-1.244,none
bmi,0.284,-0.053,none
children,0.937,0.201,log1p
charges,1.515,1.604,log1p


## #12. 최종 기술 통계량 확인

### 1. 최종 통계량 표 확인

In [ ]:
desc_df.T

,age,bmi,children,charges
count,1337.000,1337.000,1337.000,1337.000
mean,39.222,30.663,1.096,13279.121
std,14.044,6.100,1.206,12110.360
min,18.000,15.960,0.000,1121.874
25%,27.000,26.290,0.000,4746.344
50%,39.000,30.400,1.000,9386.161
75%,51.000,34.700,2.000,16657.717
max,64.000,53.130,5.000,63770.428
rel_diff,0.006,0.009,0.096,0.415
rdiff_flag,similar,similar,similar,diff


### 2. 기술 통계량 표 저장

In [ ]:
desc_df.to_excel("insurance_qtcheck_desc.xlsx", index=True)

## 💡 인사이트

**1. Charges (의료보험 청구 비용) - 주의 필요**
- 평균(13,279)과 중앙값(9,386) 차이가 큼 (상대차이율 0.415) → 비대칭 분포
- 왜도 1.515로 **강한 우측 꼬리 분포** → 고액 의료비가 우측에 집중
- 상한 이상치 139개(10.4%) → 상당한 극값 존재
- **로그 변환(log1p) 필수** 권장

**2. Children (부양 자녀 수)**
- 왜도 0.937로 우측 꼬리 분포
- 이상치는 없으나 **로그 변환 권장** (분포 정상화)

**3. Age, BMI - 건강함**
- 평균-중앙값 거의 일치 (rel_diff < 0.01)
- 대칭적 분포, 이상치 거의 없음
- 추가 처리 불필요

**4. 데이터 품질**
- 결측치 0건, 완전중복 1건 제거 완료 (1,338→1,337)
- 전체적으로 양호한 상태